src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsCache.cs

```diff
@@ -0,0 +1,130 @@
+    internal class CodeActionsCache
+    {
+        /// <summary>
+        /// Ensures we aren't making concurrent modifications to the list of cached items.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        /// <summary>
+        /// Maximum number of cached items.
+        /// </summary>
+        private readonly int _maxCacheSize = 3;
+
+        /// <summary>
+        /// Current list of cached items.
+        /// </summary>
+        private readonly List<CodeActionsCacheItem> _cachedItems = new List<CodeActionsCacheItem>();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeActionsCache()
+        {
+        }
+
+        public async Task UpdateCacheAsync(
+            Document document,
+            LSP.Range range,
+            ImmutableArray<UnifiedSuggestedActionSet> cachedSuggestedActionSets,
+            CancellationToken cancellationToken)
+        {
+            using (await _semaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // If there's a value in the cache with the same document and range we're searching for,
+                // remove and replace it with our updated value if the updated value is different.
+                var previousCachedItem = _cachedItems.Where(
+                    c => c.Document == document && c.Range.Start == range.Start && c.Range.End == range.End);
+                if (previousCachedItem.Any())
+                {
+                    var item = previousCachedItem.First();
+                    if (item.CachedSuggestedActionSets.SequenceEqual(cachedSuggestedActionSets))
+                    {
+                        return;
+                    }
+
+                    _cachedItems.Remove(item);
+                }
+                // If the cache is full, remove the oldest cached value.
+                else if (_cachedItems.Count >= _maxCacheSize)
+                {
+                    _cachedItems.RemoveAt(0);
+                }
+
+                _cachedItems.Add(new CodeActionsCacheItem(document, range, cachedSuggestedActionSets));
+            }
+        }
+
+        /// <summary>
+        /// Attempts to retrieve the cached action sets that match the given document and range.
+        /// Returns null if no match is found.
+        /// </summary>
+        public async Task<ImmutableArray<UnifiedSuggestedActionSet>?> GetCacheAsync(
+            Document document,
+            LSP.Range range,
+            CancellationToken cancellationToken)
+        {
+            using (await _semaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                foreach (var cachedItem in _cachedItems)
+                {
+                    if (document == cachedItem.Document && document.Project.Solution == cachedItem.Document.Project.Solution &&
```

Comment:
Why doesn't UpdateCache check the Solution? Is it worth a helper method to ensure these two checks are in sync?